Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveE.cpp

```diff
@@ -0,0 +1,155 @@
+#include "Utils/WarpXAlgorithmSelection.H"
+#include "FiniteDifferenceSolver.H"
+#ifdef WARPX_DIM_RZ
+    // currently works only for 3D
+#else
+#   include "FiniteDifferenceAlgorithms/CartesianYeeAlgorithm.H"
+#   include "FiniteDifferenceAlgorithms/CartesianCKCAlgorithm.H"
+#endif
+#include "Utils/WarpXConst.H"
+#include <AMReX_Gpu.H>
+#include <WarpX.H>
+
+using namespace amrex;
+
+void FiniteDifferenceSolver::MacroscopicEvolveE (
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 >& Efield,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Bfield,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Jfield,
+    amrex::Real const dt, std::unique_ptr<MacroscopicProperties> const& macroscopic_properties ) {
+
+   // Select algorithm (The choice of algorithm is a runtime option,
+   // but we compile code for each algorithm, using templates)
+#ifdef WARPX_DIM_RZ
+    amrex::Abort("currently macro E-push does not work for RZ");
+#else
+    if (m_do_nodal) {
+        amrex::Abort(" macro E-push does not work for nodal ");
+
+    } else if (m_fdtd_algo == MaxwellSolverAlgo::Yee) {
+
+        MacroscopicEvolveECartesian <CartesianYeeAlgorithm> ( Efield, Bfield, Jfield, dt,
+                                                              macroscopic_properties );
+
+    } else if (m_fdtd_algo == MaxwellSolverAlgo::CKC) {
+
+        // Note : EvolveE is the same for CKC and Yee.
+        // In the templated Yee and CKC calls, the core operations for EvolveE is the same.
+        MacroscopicEvolveECartesian <CartesianCKCAlgorithm> ( Efield, Bfield, Jfield, dt,
+                                                              macroscopic_properties );
+    } else {
+        amrex::Abort("Unknown algorithm");
+    }
+#endif
+
+}
+
+
+#ifndef WARPX_DIM_RZ
+
+template<typename T_Algo>
+void FiniteDifferenceSolver::MacroscopicEvolveECartesian (
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 >& Efield,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Bfield,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Jfield,
+    amrex::Real const dt, std::unique_ptr<MacroscopicProperties> const& macroscopic_properties ) {
+
+    const int &macroscopic_solver_algo = WarpX::macroscopic_solver_algo;
+    Real sigma = macroscopic_properties->sigma();
+    Real const mu = macroscopic_properties->mu();
+    Real const epsilon = macroscopic_properties->epsilon();
+
+    Real alpha = 0._rt;
+    Real beta  = 0._rt;
+    Real fac1 = 0._rt;
+    Real inv_fac = 0._rt;
+
+    if (macroscopic_solver_algo == 0) {
```

Comment:
Can you use the enumerated values here instead of 0 and 1?

Reply:
done.